howso.amlg

```diff
@@ -217,29 +472,191 @@
+							(conclude (call !return (assoc errors (list "Specified target_trainee already exists."))) )
+						)
+					)
+				)
+			)
+
+			(if (or source_id source_name_path)
+				(let
+					(assoc
+						source_trainee_path
+							(call_entity trainee "ResolveChildPath" (assoc
+								name_path source_name_path
+								id source_id
+							))
+					)
+					(if (~ "string" source_trainee_path)
+						(conclude (conclude (call !return (assoc errors (list source_trainee_path))) ))
+					)
+					(if (not (call_entity trainee "IsChildInternal" (assoc child_trainee_path source_trainee_path)))
+						(conclude (conclude (call !return (assoc errors (list "Specified source trainee is stored independently."))) ))
+					)
+
+					(assign (assoc
+						update_hierarchy (true)
+						source_trainee (append trainee source_trainee_path)
+					))
+
+					;if copying directly into the trainee, target_trainee will be the name of the target instead of the path, overwrite as the path to it
+					(if (~ "string" target_trainee)
+						(assign (assoc target_trainee (list trainee ".trainee_container" target_trainee) ))
+					)
+				)
+			)
+
+			;do the actual copy here
+			(declare (assoc
+				name (first (clone_entities source_trainee target_trainee))
+			))
+
+			;update the hierarchy definition if copy was a success
+			(if (and name update_hierarchy)
+				(call !update_loaded_trainee_paths_and_ids (assoc
+					trainee trainee
+					child_name (last target_trainee)
+					child_trainee_path (tail target_trainee)
+				))
+			)
+
+			(call !return (assoc payload (assoc "name" name ) ))
 		)
 
 
+	;updates hierarchy paths for loaded trainee and removes loaded trainee's id as well all its contained childrens' ids as well
+	;
+	;parameters:
+	;  trainee: name of trainee thats' doing the loading
+	;  child_name: name of loaded child trainee
+	;  child_trainee_path: path to loaded child trainee in the format of [".trainee_container" child_name ...]
+	#!update_loaded_trainee_paths_and_ids
+	(let
+		(assoc subtrainee_contained_map (retrieve_from_entity (append trainee child_trainee_path) "childTraineeIsContainedMap") )
+		(call_entity trainee "AddChildTraineePaths" (assoc
+			child_name child_name
+			child_trainee_path child_trainee_path
+		))
+
+		;null out the id of the copy
+		(assign_to_entities (append trainee child_trainee_path) (assoc traineeId (null) ))
+
+		;if the loaded trainee has its own hierarchy, add all the child trainee paths for all the children of the loaded trainee
+		(if (size subtrainee_contained_map)
+			(let
+				(assoc
+					contained_trainee_path_map (retrieve_from_entity (append trainee child_trainee_path) "containedTraineeNamePathStringToListMap")
+				)
+
+				;remove all contained id references of the loaded trainee
+				(assign_to_entities (append trainee child_trainee_path) (assoc
+					containedTraineeNamePathStringToIdMap (assoc)
+					containedTraineeIdToNamePathMap (assoc)
+					traineeId (null)
```

Comment:
this one happens just above, so is this redundant?

Reply:
no, one block is for the main trainee, the others are for any contained one if there are any contained